soccer/gameplay/tactics/advance_zone_midfielder.py

```diff
@@ -0,0 +1,174 @@
+import main
+import robocup
+import behavior
+import constants
+import enum
+import math
+
+import composite_behavior
+import skills.move
+import evaluation.ball
+import evaluation.passing_positioning
+import evaluation.passing
+import evaluation.shooting
+import functools
+import plays.offense.adaptive_formation
+
+
+#2 midfielder rely on the future location of their teammate to pass quickly
+class AdvanceZoneMidfielder(composite_behavior.CompositeBehavior):
+    # Weights for the general field positioning
+    FIELD_POS_WEIGHTS = (0.01, 3, 0.02)#AdaptiveFormation.FIELD_POS_WEIGHTS
+    # Weights for finding best pass
+    PASSING_WEIGHTS = (2, 2, 15, 10)#AdaptiveFormation.PASSING_WEIGHTS
+    # Initial arguements for the nelder mead optimization in passing positioning
+    NELDER_MEAD_ARGS = (robocup.Point(0.75, 1), \
+                        robocup.Point(0.01, 0.01), 1, 1.1, \
+                        0.5, 0.9, 100, 1, 0.1)#AdaptiveFormation.NELDER_MEAD_ARGS
+
+    class State(enum.Enum):
+        ## getting ready to recieve a pass from another robot
+        passSet = 1
+
+        hold = 2
+        
+
+    def __init__(self):
+        super().__init__(continuous=True)
+
+        #Sends one robot to best reciever point, and other to the best receiving point
+        # of the first midfielder
+        self.add_state(AdvanceZoneMidfielder.State.passSet,
+                       behavior.Behavior.State.running)
+
+        self.add_state(AdvanceZoneMidfielder.State.hold, 
+                        behavior.Behavior.State.running)
+
+        self.moves = [None, None]
+
+
+        self.priorities = [1, 2]
+
+        self.passing_point = main.ball().pos
+
+        y_temp_hold = 0.8 * self.passing_point.y
+
+        x_temp_hold = constants.Field.Width / 3
+
+        self.hold_point = [robocup.Point(x_temp_hold,y_temp_hold), robocup.Point(x_temp_hold,y_temp_hold)]
+
+        self.dribble_state = True
+
+        self.kick = False
+
+        for s in AdvanceZoneMidfielder.State:
+            self.add_state(s, behavior.Behavior.State.running)
+
+        self.names = ['left','right']
+        #intially should be passSet method
+        self.add_transition(behavior.Behavior.State.start,
+                            AdvanceZoneMidfielder.State.passSet, lambda: True,
+                            "Immediately")
+        self.add_transition(AdvanceZoneMidfielder.State.passSet,
+                            AdvanceZoneMidfielder.State.hold, lambda: False,
```

Comment:
I assume this is a holdover from testing and should actually be using some real lambda function here. Otherwise it will only be in the passSet state

Reply:
I switched it over to actually holding during a kick. I'm not sure this is a desired behavior. But this should prevent us going behind the goal during a kick and getting in the way of the kick.